src/Database/LSMTree/Internal.hs

```diff
@@ -705,11 +703,13 @@ updates resolve es th = do
 retrieveBlobs ::
      m ~ IO  -- TODO: replace by @io-classes@ constraints for IO simulation.
   => Session m h
-  -> V.Vector (BlobRef m (FS.Handle h))
+  -> V.Vector (WeakBlobRef m (FS.Handle h))
   -> m (V.Vector SerialisedBlob)
-retrieveBlobs sesh refs =
+retrieveBlobs sesh wrefs =
     withOpenSession sesh $ \seshEnv ->
-      bracket_ upgradeWeakReferences removeReferences $ do
+      handle (\(BlobRef.DeRefWeakBlobRef i) -> throwIO (ErrBlobRefInvalid i)) $
+      bracket (BlobRef.deRefWeakBlobRefs wrefs)
+              (BlobRef.removeReferences) $ \refs -> do
```

Comment:
Optionally, expose this `bracket` version of dereferencing from the `BlobRef` module instead of `deRefWeakBlobRefs`, so that you are forced to use the exception safe version

Reply:
I agree, `withDeRefBlobRefs :: Vector WeakBlobRef -> (Vector BlobRef -> m r) -> m r` interface is usable here, so better to provide it directly.

Reply:
Ok, done.